wdns/rdata_to_ubuf.c

```diff
@@ -124,15 +124,32 @@ svcparam_to_str(uint16_t key, const uint8_t *src, uint16_t len, ubuf *u)
 		 * SvcParamValue; otherwise, the SvcParamValue is malformed.
 		 */
 		while ((ptr - src) < len) {
+			uint8_t l;
+
 			oclen = *ptr;
 			ptr += 1;	/* skip the length */
 
 			if ((ptr + oclen - src) > len) {
 				return (wdns_res_parse_error);
 			}
 
-			(void) rdata_to_str_string_unquoted(ptr, oclen, u);
-			ptr += oclen;
+			l = oclen;
+
+			while (l--) {
+				uint8_t c = *ptr++;
+
+				if (c == '"') {
+				        ubuf_add_cstr(u, "\\\"");
+				} else if (c == '\\') {
+				        ubuf_add_cstr(u, "\\\\");
+				} else if (c == ',') {
+					ubuf_add_fmt(u, "\\,");
+				} else if (c >= ' ' && c <= '~') {
```

Comment:
If `c == ' '` here, the output will trip up the `strtok_r(NULL, " \t\r\n", &endp)` parsing of the value. It would be more robust here to:
1. Always output a double-quoted (and escaped where necessary) string for `alpn` and default `keyNNN` values. 
2. Parse `alpn` and `keyNNN` values as possibly quoted character-string values (e.g., with `rdata_from_str_string()` into a separate ubuf, relying on `rdata_from_str_string()` to determine the end of the value)
3. For alpn values, parse the comma separated list of alpn values  from the result of (2)

The ipv[46]hint, port, and ech values can be (and are) encoded with a limited character set not requiring quoting, regardless of their wire format contents, which is why the standard can (and does) prohibit escape characters in them. I read the standard as allowing these to be quoted, but handling that is not strictly necessary. I expect wdns to be able to parse its own output back to the original rdata, and for that output to be parseable by other DNS zone file implementations, but do not expect wdns to parse all valid zone file contents.

Reply:
Ok.. I think I addressed your comments. I'm wrapping alpn and keyNNN values around double quotes (in presentation format only). Please take a look.

Reply:
The presentation format looks right with these latest changes, but the parsing needs work.  In particular, it will fail if any service parameter value contains a space.  The following patch to the rdata_to_str (+ round trip) test cases illustrates the remaining. This will require moving away from `strtok_r` on whitespace to separate service parameters, as that cannot account for quoted spaces, which will involve a larger refactor of the parsing.

`
--- a/t/test-rdata_to_str.c
+++ b/t/test-rdata_to_str.c
@@ -410,13 +410,13 @@ struct test tdata[] = {
                    "\x00\x01"          /* alpn in network order */
                    "\x00\x03"          /* length of the alpnid in net order */
                    "\x02\x68\x32"      /* length-value */
-                   "\x00\x09"          /* key9 in network order */
-                   "\x00\x03"          /* length of key9 in net order */
-                   "\x61\x6e\x79",     /* length-value */
-               .input_len = 17,
+                   "\x00\x08"          /* key8 in network order */
+                   "\x00\x09"          /* length of key9 in net order */
+                   "has space",        /* length-value */
+               .input_len = 23,
                .rrclass = WDNS_CLASS_IN,
                .rrtype = WDNS_TYPE_HTTPS,
-               .expected = "1 . alpn=\"h2\" key9=\"any\"",
+               .expected = "1 . alpn=\"h2\" key8=\"has space\"",
        },
 
        { 0 }
`

Reply:
Ok, I re-wrote the parsing of alpn and arbitrary key values to account for escaped commas and blank spaces. Please take a look.